pallets/watchtower/src/lib.rs

```diff
@@ -0,0 +1,1195 @@
+                        "[VALIDATE_UNSIGNED] REJECTED (Signature Invalid): Invalid signature for node {:?}.",
+                        node
+                    );
+                    InvalidTransaction::BadSigner.into()
+                }
+            } else {
+                log::warn!(
+                    target: "runtime::watchtower::validate",
+                    "[VALIDATE_UNSIGNED] WARNING: Received non-watchtower-vote call for unsigned validation: {:?}.",
+                    call
+                );
+                InvalidTransaction::Call.into()
+            }
+        }
+    }
+
+    impl<T: Config> Pallet<T> {
+        pub fn notify_summary_ready_for_validation(
+            instance: SummarySourceInstance,
+            root_id: WatchtowerRootId<BlockNumberFor<T>>,
+            root_hash: WatchtowerOnChainHash,
+        ) -> DispatchResult {
+            log::info!(
+                target: "runtime::watchtower::notification",
+                "Received notification: instance={:?}, root_id={:?}, hash={:?}",
+                instance, root_id, root_hash
+            );
+
+            let consensus_key = (instance, root_id.clone());
+
+            if VoteConsensusReached::<T>::get(&consensus_key) {
+                log::warn!(
+                    target: "runtime::watchtower::notification",
+                    "Consensus already reached for summary: {:?}",
+                    consensus_key
+                );
+                return Ok(());
+            }
+
+            if VotingStartBlock::<T>::get(&consensus_key).is_some() {
+                log::warn!(
+                    target: "runtime::watchtower::notification",
+                    "Voting already active for summary: {:?}",
+                    consensus_key
+                );
+                return Ok(());
+            }
+
+            if root_hash == sp_core::H256::zero() {
+                log::warn!(
+                    target: "runtime::watchtower::notification",
+                    "Received empty root hash for summary: {:?}",
+                    consensus_key
+                );
+            }
+
+            VotingStartBlock::<T>::insert(
+                &consensus_key,
+                frame_system::Pallet::<T>::block_number(),
+            );
+            PendingValidationRootHash::<T>::insert(&consensus_key, root_hash);
+
+            log::info!(
+                target: "runtime::watchtower::notification",
+                "Started voting period for summary: instance={:?}, root_id={:?}, hash={:?}",
+                instance, root_id, root_hash
+            );
+
+            Ok(())
+        }
+
+        fn process_pending_validations() {
+            log::debug!(target: "runtime::watchtower::ocw", "Processing pending validations");
+
+            let current_block = frame_system::Pallet::<T>::block_number();
+            let voting_period = VotingPeriod::<T>::get();
+
+            for (consensus_key, start_block) in VotingStartBlock::<T>::iter() {
```

Comment:
Wouldn't it be better to apply the filtering when reading from `VotingStartBlock` to exclude those that have reached consensus?